app/(gcforms)/[locale]/(app administration)/admin/(with nav)/events/actions.ts

```diff
@@ -0,0 +1,236 @@
+          UserID: event.UserID,
+          "Event#SubjectID#TimeStamp": event["Event#SubjectID#TimeStamp"],
+        })),
+      },
+    },
+  });
+
+  await dynamoDBDocumentClient.send(batchRequest).then(async (data) => {
+    auditLogs.push(
+      ...(data?.Responses?.AuditLogs?.map((item) => ({
+        UserID: item.UserID,
+        Event: item.Event,
+        TimeStamp: item.TimeStamp,
+        Description: item.Description,
+        Subject: item.Subject,
+      })) ?? [])
+    );
+
+    if (data.UnprocessedKeys?.AuditLogs) {
+      while (retries < maxRetries) {
+        // eslint-disable-next-line no-await-in-loop
+        await delay(200); // Wait for 200ms second before retrying
+        const retryRequest = new BatchGetCommand({
+          RequestItems: {
+            AuditLogs: {
+              Keys: data.UnprocessedKeys.AuditLogs.Keys,
+            },
+          },
+        });
+        // eslint-disable-next-line no-await-in-loop
+        const retryResponse = await dynamoDBDocumentClient.send(retryRequest);
+        auditLogs.push(
+          ...(retryResponse.Responses?.AuditLogs.map((item) => ({
+            UserID: item.UserID,
+            Event: item.Event,
+            TimeStamp: item.TimeStamp,
+            Description: item.Description,
+            Subject: item.Subject,
+          })) ?? [])
+        );
+        if (!retryResponse.UnprocessedKeys?.AuditLogs) {
+          break; // Exit the loop if there are no more unprocessed keys
+        }
+        retries++;
+      }
+    }
+  });
+  return auditLogs;
+};
+
+export const getEventsForUser = async (userId: string) => {
+  const {
+    user: { id: callingUserId },
+  } = await authorization.canViewAllUsers();
+  const events = await _retrieveEvents({
+    TableName: "AuditLogs",
+    IndexName: "UserByTime",
+    Limit: 100,
+    KeyConditionExpression: "UserID = :userId",
+    ExpressionAttributeValues: {
+      ":userId": userId,
+    },
+    ScanIndexForward: false,
+  });
+  logEvent(
+    callingUserId,
+    { type: "User", id: userId },
+    "AuditLogsRead",
+    `User ${callingUserId} read audit logs for user ${userId}`
+  );
+  return events;
+};
+
+export const getEventsForForm = async (formId: string) => {
+  const {
+    user: { id: callingUserId },
+  } = await authorization.canViewAllForms();
+
+  const events = await _retrieveEvents({
+    TableName: "AuditLogs",
+    IndexName: "SubjectByTimestamp",
+    Limit: 100,
+    KeyConditionExpression: "Subject = :formId",
+    ExpressionAttributeValues: {
+      ":formId": `Form#${formId}`,
+    },
+    ScanIndexForward: false,
+  });
+  logEvent(
+    callingUserId,
+    { type: "Form", id: formId },
+    "AuditLogsRead",
+    `User ${callingUserId} read audit logs for form ${formId}`
+  );
+  return events;
+};
+
+const emailSchema = object({
+  subject: string([email()]),
+});
+
+const formSchema = object({
+  subject: string([minLength(1), maxLength(30), toLowerCase(), toTrimmed()]),
```

Comment:
Unless there is something I did not understand. An email address with more than 30 characters would not be accepted. We may want to increase the limit?

Reply:
```suggestion
  subject: string([minLength(1), maxLength(100), toLowerCase(), toTrimmed()]),
```

Reply:
Right, I was thinking more about the CUID and not the email.